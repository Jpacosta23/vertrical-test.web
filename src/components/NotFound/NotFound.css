#notfound {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-direction: column;
    height: 100vh;
    margin-top: 50px;
}

.notfound-404 h1 {
    font-family: "Montserrat", sans-serif;
    text-align: center;
    font-size: 220px;
    font-weight: 900;
    margin: 10px;
    color: #2f81ff;
}

#notfound h2 {
    text-align: center;
    font-size: 22px;
    font-weight: 700;
    text-transform: uppercase;
    color: black;
    margin-top: 20px;
    margin-bottom: 15px;
}
