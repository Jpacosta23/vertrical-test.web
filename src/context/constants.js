export const GET_PHOTOS = "get-photos";
export const SET_LOADING = "set-loading";
export const GET_PHOTO = "get-photo";
export const DELETE_PHOTO = "delete-photo";
