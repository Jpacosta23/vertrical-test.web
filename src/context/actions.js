import { SET_LOADING, GET_PHOTOS, GET_PHOTO } from "./constants";
import {
    getAllPhotos,
    getPhotoById,
    getPhotosByTitle,
} from "../services/photo.service";

export async function getPhotos(dispatch) {
    dispatch({ type: SET_LOADING, payload: true });
    try {
        const photos = await getAllPhotos();
        dispatch({ type: GET_PHOTOS, payload: photos });
    } catch (error) {
        throw Error(error);
    } finally {
        dispatch({ type: SET_LOADING, payload: false });
    }
}

export async function getPhotosById(dispatch, id) {
    dispatch({ type: SET_LOADING, payload: true });
    try {
        const photo = await getPhotoById(id);
        dispatch({ type: GET_PHOTO, payload: photo });
    } catch (error) {
        throw Error(error);
    } finally {
        dispatch({ type: SET_LOADING, payload: false });
    }
}

export async function getPhotosFilteredByTitle(dispatch, title) {
    dispatch({ type: SET_LOADING, payload: true });
    try {
        const photos = await getPhotosByTitle(title);
        dispatch({ type: GET_PHOTOS, payload: photos });
    } catch (error) {
        throw Error(error);
    } finally {
        dispatch({ type: SET_LOADING, payload: false });
    }
}
