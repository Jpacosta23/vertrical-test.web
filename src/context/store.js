import React, { createContext, useReducer, useContext } from 'react';

import { GET_PHOTOS, SET_LOADING, GET_PHOTO, DELETE_PHOTO } from './constants';

const AppStateContext = createContext();
const AppDispatchContext = createContext();

const initialState = {
  photos: [],
  currentPhoto: {},
  isLoading: false
};

function AppReducer(state, action) {
  switch (action.type) {
    case GET_PHOTOS: {
      return {
        ...state,
        photos: action.payload
      };
    }
    case GET_PHOTO: {
      return {
        ...state,
        currentPhoto: action.payload
      };
    }
    case DELETE_PHOTO: {
      return {
        ...state,
        currentPhoto: action.payload
      };
    }
    case SET_LOADING: {
      return {
        ...state,
        loading: action.payload
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
}

/* eslint-disable */
function AppProvider({ children }) {
  const [state, dispatch] = useReducer(AppReducer, initialState);
  return (
    <AppStateContext.Provider value={state}>
      <AppDispatchContext.Provider value={dispatch}>{children}</AppDispatchContext.Provider>
    </AppStateContext.Provider>
  );
}

function useAppState() {
  const context = useContext(AppStateContext);
  if (context === undefined) {
    throw new Error('useAppState must be used within a AppProvider');
  }
  return context;
}

function useAppDispatch() {
  const context = useContext(AppDispatchContext);
  if (context === undefined) {
    throw new Error('useAppDispatch must be used within a AppProvider');
  }
  return context;
}

export { AppProvider, useAppState, useAppDispatch };
