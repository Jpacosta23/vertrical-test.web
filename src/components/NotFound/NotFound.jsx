import React from 'react';
import { useNavigate } from 'react-router-dom';
import { deleteCurrentPhoto } from '../../context/actions';
import { useAppDispatch, useAppState } from '../../context/store';
import PropTypes from 'prop-types';
import './NotFound.css';
const NotFound = ({ status, error, showButton = true }) => {
  const { isLoading } = useAppState();
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const handleReturn = () => {
    deleteCurrentPhoto(dispatch);
    navigate('/');
  };
  return (
    <>
      {!isLoading ? (
        <div className="container">
          <div id="notfound">
            <div className="notfound-404">
              <h1>{status}</h1>
            </div>
            <h2>{error}</h2>
            {showButton ? (
              <button id="exit-button" onClick={handleReturn} data-testid="button">
                <p>Go Home</p>
              </button>
            ) : null}
          </div>
        </div>
      ) : null}
    </>
  );
};

NotFound.propTypes = {
  status: PropTypes.string,
  error: PropTypes.string,
  showButton: PropTypes.bool
};

export default NotFound;
