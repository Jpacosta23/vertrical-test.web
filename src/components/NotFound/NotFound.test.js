import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import NotFound from './NotFound';
import renderer from 'react-test-renderer';
import { AppProvider } from '../../context/store';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate // Return an empty jest function to test whether it was called or not...I'm not depending on the results so no need to put in a return value
}));

afterEach(cleanup);
describe('Not Found component test', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <AppProvider>
        <MemoryRouter>
          <NotFound />
        </MemoryRouter>
      </AppProvider>,
      div
    );
  });
  it('goes to the home page when button clicked', () => {
    const { container } = render(
      <AppProvider>
        <MemoryRouter>
          <NotFound />
        </MemoryRouter>
      </AppProvider>
    );
    const buttonElement = screen.getByRole('button');
    fireEvent.click(buttonElement);
    expect(mockedUsedNavigate).toHaveBeenCalledTimes(1);
  });
  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <AppProvider>
          <MemoryRouter>
            <NotFound />
          </MemoryRouter>
        </AppProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
