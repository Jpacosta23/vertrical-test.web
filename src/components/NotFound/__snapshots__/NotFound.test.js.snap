// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not Found component test matches snapshot 1`] = `
<div
  className="container"
>
  <div
    id="notfound"
  >
    <div
      className="notfound-404"
    >
      <h1 />
    </div>
    <h2 />
    <button
      data-testid="button"
      id="exit-button"
      onClick={[Function]}
    >
      <p>
        Go Home
      </p>
    </button>
  </div>
</div>
`;
